Library/Homebrew/formula_installer.rb

```diff
@@ -457,12 +457,13 @@ def expand_requirements
           Requirement.prune
         elsif req.satisfied?
           Requirement.prune
-        elsif include_test? && req.test?
-          next
-        elsif !runtime_requirements.include?(req) && install_bottle_for_dependent
-          Requirement.prune
-        elsif (dep = formula_deps_map[dependent.name]) && dep.build?
-          Requirement.prune
+        elsif req.test? || req.build?
+          keep = false
+          keep ||= runtime_requirements.include?(req)
+          keep ||= req.test? && include_test? && dependent == f
+          keep ||= req.build? && !install_bottle_for_dependent
+          keep ||= (dep = formula_deps_map[dependent.name]) && !dep.build?
+          Requirement.prune unless keep
```

Comment:
In the case `keep` is true, I think we want to fallthrough to the `else` case. At this point, the requirment has already been established as not satisfied (line 458).

Reply:
Yeh, good call. Pushed a tweak.